src/Features/Core/Portable/NavigateTo/IRemoteNavigateToSearchService.cs

```diff
@@ -2,19 +2,63 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
 using System.Collections.Immutable;
+using System.Composition;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.Remote;
 
 namespace Microsoft.CodeAnalysis.NavigateTo
 {
     internal interface IRemoteNavigateToSearchService
     {
-        ValueTask<ImmutableArray<SerializableNavigateToSearchResult>> SearchDocumentAsync(
-            PinnedSolutionInfo solutionInfo, DocumentId documentId, string searchPattern, ImmutableArray<string> kinds, CancellationToken cancellationToken);
+        ValueTask SearchDocumentAsync(
+            PinnedSolutionInfo solutionInfo, DocumentId documentId, string searchPattern, ImmutableArray<string> kinds, RemoteServiceCallbackId callbackId, CancellationToken cancellationToken);
 
-        ValueTask<ImmutableArray<SerializableNavigateToSearchResult>> SearchProjectAsync(
-            PinnedSolutionInfo solutionInfo, ProjectId projectId, ImmutableArray<DocumentId> priorityDocumentIds, string searchPattern, ImmutableArray<string> kinds, CancellationToken cancellationToken);
+        ValueTask SearchProjectAsync(
+            PinnedSolutionInfo solutionInfo, ProjectId projectId, ImmutableArray<DocumentId> priorityDocumentIds, string searchPattern, ImmutableArray<string> kinds, RemoteServiceCallbackId callbackId, CancellationToken cancellationToken);
+
+        public interface ICallback
+        {
+            ValueTask OnResultFoundAsync(RemoteServiceCallbackId callbackId, SerializableNavigateToSearchResult result);
+        }
+    }
+
+    [ExportRemoteServiceCallbackDispatcher(typeof(IRemoteNavigateToSearchService)), Shared]
+    internal sealed class NavigateToSearchServiceServerCallbackDispatcher : RemoteServiceCallbackDispatcher, IRemoteNavigateToSearchService.ICallback
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public NavigateToSearchServiceServerCallbackDispatcher()
+        {
+        }
+
+        private new NavigateToSearchServiceCallback GetCallback(RemoteServiceCallbackId callbackId)
+            => (NavigateToSearchServiceCallback)base.GetCallback(callbackId);
+
+        public ValueTask OnResultFoundAsync(RemoteServiceCallbackId callbackId, SerializableNavigateToSearchResult result)
+            => GetCallback(callbackId).OnResultFoundAsync(result);
+    }
+
+    internal sealed class NavigateToSearchServiceCallback
```

Comment:
callbacks for oop seem a little complicated.  It's probably been considered, but why can't the logic for retrieving the actual callback method be abstracted and 'pre-processed' before any request hits the IRemoteXXXService? That they don't need to do hop from callbackid -> callback in each implementation.  Is it because of the rehydrating stuff that is going on here that requires the solution (that other requests don't need?)

Reply:
not being familiar with the infrastructure, it was fairly difficult to follow the chain

Reply:
i don't have the answer to that.  and, fwiw, i agree with you and think that would be reasonable.  @tmat for his thoughts here.  I was just following an existin gpattern that we already have in FindRefs.  It does seem like we could potentially simplify things here in some way.